Best Apps for App Store and Google Play Markets

In this project we look at apps available on the App Store and Google Play applications. In analyzing the data, we will determine which free apps are most effective to use in getting more customers.We Will only consider free apps as all of our revenue will be generated via in-ad sales. Let's get started!

In [1]:
AppStore_file = open('AppleStore.csv')
from csv import reader
AppStore_read_file = reader(AppStore_file)
AppStore_data = list(AppStore_read_file)
AppStore_data_header = AppStore_data[0]
AppStore_data  = AppStore_data[1:]

def explore_dataset(dataset,start,end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
AppleStore = explore_dataset(AppStore_data,0,2,True)        

GooglePlay_file = open('googleplaystore.csv')
GooglePlay_read_file = reader(GooglePlay_file)
GooglePlay_data = list(GooglePlay_read_file)
GooglePlay_data_header = GooglePlay_data[0]
GooglePlay_data = GooglePlay_data[1:]

GooglePlay = explore_dataset(GooglePlay_data, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7197
Number of columns: 16
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In [2]:
print("Apple Store App Names")
print(AppStore_data_header)
print('\n')
print("Google Play App Names")
print(GooglePlay_data_header)

Apple Store App Names
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Google Play App Names
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [3]:
print("These columns could help us in analysis of the best apps in the Apple App Store: price, user rating, prime genre, and size bytes.")
print("In Google Play, the features listed are useful for determining the best apps. It's helpful to look at the data in more detail to determine which ones are most important.")

These columns could help us in analysis of the best apps in the Apple App Store: price, user rating, prime genre, and size bytes.
In Google Play, the features listed are useful for determining the best apps. It's helpful to look at the data in more detail to determine which ones are most important.


For detailed info on the descriptors for the Apple Store applications please see [here](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps).
And for more info on the descriptors used in the Google Play dataset, see the [Google Play mobile app store](https://www.kaggle.com/lava18/google-play-store-apps).

In [4]:
print(GooglePlay_data_header)
print('\n')
print(GooglePlay_data[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [5]:
del GooglePlay_data[10472]

The GooglePlay data was missing the Life Made WI-Fi Touchscreen Photo Frame app rating and thus the app was removed from our dataset in the preceding code. But another problem exists. GooglePlay also has duplicate entries which we need to remove. Let's see what these are.

In [6]:
duplicate_apps = []
unique_apps = []

for app in GooglePlay_data:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:17])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling']


For the sake of accuracy in our analysis, we will be removing the duplicate entries. These will not be deleted randomly; we will use the number of reviews to decide which entry to keep. The higher the number of reviews, the more recent the data. Hence, we will keep the entry with the highest number of reviews for all duplicated entries.

In [7]:
reviews_max = {}
for app in GooglePlay_data:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    if name not in reviews_max:
        reviews_max[name] = n_reviews

print("Number of Google Play apps is ", len(reviews_max))     

Number of Google Play apps is  9659


In [8]:
android_clean = []
already_added = []
for app in GooglePlay_data:
    name = app[0]
    n_reviews = float(app[3])
#Check that the review in this row matches the highest review#
#obtained and the app has not already been added (accounted for)#    
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        android_clean.append(app)        
        already_added.append(name)

explore_dataset(android_clean,0,2,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9659
Number of columns: 13


As expected, we have 9659 Google Play apps.

The Engl function below is meant to return False if an app name falls outside of the ASCII English characters range.

In [9]:
def Engl(string):
    for character in string:
        
        if ord(character) > 127:
            return False
    return True
    
print(Engl('Instagram'))
print(Engl('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(Engl('Docs To Go™ Free Office Suite'))
print(Engl('Instachat 😜'))

True
False
False
False


As you can see, Docs To Go™ Free Office Suite and Instachat 😜return False although they are English apps. This is because of the ™ symbol and the 😜 emoji, which have ASCII numbers greater than 127. To avoid leaving out Docs To Go™ Free Office Suite and Instachat 😜, as well as other English apps with special characters, we'll refine our Engl function.

In [10]:
def English(string):
    ascii_out = 0
    for character in string:
        
        if ord(character) > 127:
            ascii_out += 1
    if ascii_out > 3:    
        return False
    return True

print(English('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(English('Docs To Go™ Free Office Suite'))
print(English('Instachat 😜'))


False
True
True


In [11]:
GooglePlay_English = []
AppleStore_English = []
for app in android_clean:
    name = app[0]
    if English(name):
        GooglePlay_English.append(app)
        
for app in AppStore_data:
    name = app[1]
    if English(name):
        AppleStore_English.append(app)
        
print('English GooglePlay apps')
explore_dataset(GooglePlay_English, 0, 3, True)
print('\n')
print('English AppleStore apps')
explore_dataset(AppleStore_English, 0, 3, True)    
    

English GooglePlay apps
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


English AppleStore apps
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '

So, after removing duplicate and nonEnglish apps, we're left with 9614 GooglePlay English apps and 6183 English AppleStore apps.


Now, we want to isolate the free apps in a separate list. Because our revenue will be generated via in-app ads, we are looking at free apps only.

In [12]:
print("Apple Store App Columns") 
print(AppStore_data_header)
print('\n')
print("Google Play App Names")
print(GooglePlay_data_header)

Apple Store App Columns
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Google Play App Names
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The price for the Apple Store apps is index 4. And for Google Play it is index 7.

In [13]:
GooglePlay_EnglishFree = []
AppleStore_EnglishFree = []
for app in GooglePlay_English:
    price = app[7]
    if price == '0':
        GooglePlay_EnglishFree.append(app)
        
for app in AppleStore_English:
    price = app[4]
    if price == '0.0':
        AppleStore_EnglishFree.append(app)
        
print('Free English GooglePlay apps')
explore_dataset(GooglePlay_EnglishFree, 0, 3, True)
print('\n')
print('Free English AppleStore apps')
explore_dataset(AppleStore_EnglishFree, 0, 3, True)  

Free English GooglePlay apps
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8864
Number of columns: 13


Free English AppleStore apps
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD'

We want an app that can be used on GooglePlay and the App Store. At the same time, we want to minimize risk and overhead.To achieve these goals, we will initially build a minimal version of the app for use in Google Play. If the response is good then we'll further develop the app. After six months, if the app is profitable, we will produce an iOS version for use in the App Store. Since we want to use the app in both GooglePlay and the App Store, we'll need to determine what genres these two services have in common. Let's look at their column names.

In [14]:
print("Apple Store App Columns") 
print(AppStore_data_header)
print('\n')
print("Google Play App Names")
print(GooglePlay_data_header)
      

Apple Store App Columns
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Google Play App Names
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The Apple Store : Google Play apps that the two services have in common are: 
1. track name: app
2. size bytes : size
3. price : price
4. user rating : rating
5. user rating count : reviews
6. prime genre : genres
7. ver : current ver
8. cont rating : content rating
9. category in Google Play can also be used to classify the type of apps.

We would compare prime genre and genre (#6 above), as well as category (#9) to create frequency tables showing the most common genres in each market.




In [15]:
def freq_table(dataset, index):
    table = {}
    total = 0
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentages = {}
    for key in table:
        percentage = (table[key]/total) * 100
        table_percentages[key] = percentage
    return table_percentages
    
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Below is the frequency table for the prime genre column in the Apple Store dataset.

In [16]:
display_table(AppleStore_EnglishFree, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


As we can see above, games accounts for almost 60% of the free English Apple Store apps. The runner up is entertainment which accounts for roughly 8% of the apps. In general, work related functions such as finance, business, and productivity are less than 2%. Most of the free English apps are for entertainment purposes. We need more info in order to recommend an app profile for the App Store. If there's a large number of apps for a particular genre this only shows what's available. Availability does not imply demand i.e. that apps of that genre have a large number of users.

Below is the category frequency table for Google Play.

In [17]:
display_table(GooglePlay_EnglishFree, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

Family, game, and tools are the top categories. In Google Play, we see that family consists of games for young children, so essentially the top two categories are games.

And finally the genres frequencey tale for Google Play:

In [18]:
display_table(GooglePlay_EnglishFree, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

Tools is in the top 3 for the category and genres columns in the Google Play data set. As in the App Store, entertainment is a frequent genre in Google Play -although it is not a popular category. Google Play has more practical apps whereas the Apple Store has mostly entertainment apps. Again, these tables only show what is available, they do not inicate what kinds of apps users prefer.

In [19]:
genre_AppStore = freq_table(AppleStore_EnglishFree, -5)

for genre in genre_AppStore:
    total = 0
    len_genre = 0
    for app in AppleStore_EnglishFree:
        genre_app = app[-5]
        if genre_app == genre:
            number_ratings = float(app[5]) #rating count total
            total = total + number_ratings
            len_genre += 1
                    
    avg_number_ratings = total/len_genre
    print(genre, ':', avg_number_ratings)

Catalogs : 4004.0
Sports : 23008.898550724636
Social Networking : 71548.34905660378
Entertainment : 14029.830708661417
Productivity : 21028.410714285714
Shopping : 26919.690476190477
Games : 22788.6696905016
Health & Fitness : 23298.015384615384
Food & Drink : 33333.92307692308
Lifestyle : 16485.764705882353
Finance : 31467.944444444445
Travel : 28243.8
Business : 7491.117647058823
Utilities : 18684.456790123455
News : 21248.023255813954
Education : 7003.983050847458
Navigation : 86090.33333333333
Photo & Video : 28441.54375
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Medical : 612.0
Reference : 74942.11111111111


Interestingly, navigation apps had the highest number of ratings of all the Apple Store apps. These were followed by reference and social networking genres. Even though the App Store favors fun free English apps, the high number of ratings for these particular genres suggest that users value practicality over fun when choosing free English apps.

In [24]:
genre_GooglePlay = freq_table(GooglePlay_EnglishFree, -4)

for category in genre_GooglePlay:
    total = 0
    len_category = 0
    number_installs = 0
    for app in GooglePlay_EnglishFree:
        category_app = app[-4]
        if category_app == category:
            number_installs = app[5]
            number_installs = number_installs.replace('+', '')
            number_installs = number_installs.replace(',', '')
            n_installs = float(number_installs)
            total += n_installs
            len_category += 1
            
    avg_n_installs = total/len_category
    print(category,":", avg_n_installs)                                
            
        

Video Players & Editors;Music & Video : 7500000.0
Simulation : 3475484.08839779
Trivia : 3475712.7027027025
Puzzle : 8302861.91
Health & Fitness;Action & Adventure : 1000000.0
Art & Design : 2122850.9433962265
Health & Fitness;Education : 100000.0
News & Magazines : 9549178.467741935
House & Home : 1331540.5616438356
Casual;Brain Games : 1425916.6666666667
Trivia;Education : 100.0
Educational;Education : 1737143.142857143
Entertainment;Music & Video : 6413333.333333333
Entertainment;Brain Games : 3314285.714285714
Adventure;Education : 10000000.0
Casual;Education : 1000000.0
Strategy;Education : 500000.0
Entertainment;Creativity : 4000000.0
Casual;Action & Adventure : 12916666.666666666
Travel & Local : 14051476.145631067
Education;Creativity : 2875000.0
Music : 9445583.333333334
Productivity : 16787331.344927534
Medical : 120550.61980830671
Music & Audio;Music & Video : 500000.0
Dating : 854028.8303030303
Role Playing;Action & Adventure : 7000000.0
Board : 4759209.117647059
Action : 1

Above is the app genre to average number of installs for the Google Play free English apps. 

Education and action & adventure had high average installs.The Apple Store findings showed that users rated practical apps more times than fun ones. Navigation apps had the highest number of ratings for Apple Store free English apps. These apps educate the customer on how to reach their destination (i.e. an adventure). My recommendation is that free English Apps should be educational and provide travel related information. 